File: zanshin/services/ticket_service.py
"""Opening a tracker ticket for the problems that would fail a build.

SARIF closed the loop towards the developer: a finding now appears on the pull request
that introduced it. This closes it towards the organisation — a problem that nobody is
going to fix this afternoon needs to exist somewhere people plan work, not only in a
dashboard nobody has a reason to open.

Design decisions, in order of how much they shaped the code:

* **Driven by the gate policy, not by a second threshold.** "Open a ticket for what
  would fail a build" is a rule an operator can already reason about, and it means
  there is exactly one place where "serious enough to act on" is defined. Inventing a
  `ticket_min_severity` would have created two vocabularies that drift apart, and the
  first bug report would be "why did it ticket that but not fail the build".
* **A sweep, not an event.** Notifications fire inline after a scan; tickets do not.
  A sweep over "actionable issues with no ticket" is idempotent by construction — the
  reference stored on the issue *is* the deduplication key — so a tracker that was in
  maintenance gets retried on the next tick instead of losing the ticket silently.
  This is also why no outbox is needed here: the state to reconcile is already in the
  issue row.
* **One ticket per issue, for the issue's whole life.** Not per scan, and not
  reopened when the issue reappears: a tracker ticket that comes back from the dead on
  every rescan is how people learn to mute the project. `Issue.ticket_ref` is set once
  and never cleared.
* **The token is encrypted at rest.** It grants write access to the tracker, which is
  a different class of secret from a webhook URL, so it goes through
  `EncryptionService` like an SSH key rather than sitting in the settings table in
  clear.
"""
import logging
from typing import Any, Dict, List, NamedTuple, Optional

import httpx

from zanshin.models.issue import Issue
from zanshin.services.encryption_service import EncryptionService
from zanshin.services.settings_service import SettingsService
from zanshin.services.url_guard import UnsafeUrlError, validate_outbound_url

logger = logging.getLogger(__name__)

PROVIDER_NONE = "none"
PROVIDER_GITLAB = "gitlab"
PROVIDER_JIRA = "jira"
VALID_PROVIDERS = (PROVIDER_NONE, PROVIDER_GITLAB, PROVIDER_JIRA)

SETTING_KEY_PROVIDER = "ticket_provider"
SETTING_KEY_BASE_URL = "ticket_base_url"
SETTING_KEY_PROJECT = "ticket_project"
SETTING_KEY_TOKEN = "ticket_token"
# Jira needs the account e-mail alongside the API token for basic auth; GitLab does
# not use it.
SETTING_KEY_USER = "ticket_user"
SETTING_KEY_ISSUE_TYPE = "ticket_issue_type"
SETTING_KEY_LABELS = "ticket_labels"
# Self-hosted GitLab and Jira are routinely on a private network, so unlike the
# notification webhook this defaults to *allowing* it — with the setting still there
# to forbid it for a deployment that only uses a hosted tracker.
SETTING_KEY_ALLOW_PRIVATE = "ticket_allow_private_url"

TOKEN_CONTEXT = "setting:ticket_token"
DEFAULT_JIRA_ISSUE_TYPE = "Bug"
DEFAULT_LABELS = "zanshin,security"
HTTP_TIMEOUT_SECONDS = 15.0

# Ceiling per sweep. A first run against a mature backlog would otherwise open several
# hundred tickets in one burst, which is both a rate-limit problem and a social one.
MAX_TICKETS_PER_SWEEP = 20


class Ticket(NamedTuple):
    reference: str
    url: str


class TicketService:
    def __init__(
        self,
        settings_service: SettingsService,
        encryption_service: Optional[EncryptionService] = None,
        http_post=httpx.post,
    ):
        self.settings_service = settings_service
        self.encryption_service = encryption_service
        self._http_post = http_post

    # --- Configuration ---

    def provider(self) -> str:
        value = (self.settings_service.get_setting(SETTING_KEY_PROVIDER, PROVIDER_NONE) or "").strip().lower()
        return value if value in VALID_PROVIDERS else PROVIDER_NONE

    def is_enabled(self) -> bool:
        return (
            self.provider() != PROVIDER_NONE
            and bool(self.base_url())
            and bool(self.project())
            and bool(self.token())
        )

    def base_url(self) -> str:
        return (self.settings_service.get_setting(SETTING_KEY_BASE_URL, "") or "").strip().rstrip("/")

    def project(self) -> str:
        return (self.settings_service.get_setting(SETTING_KEY_PROJECT, "") or "").strip()

    def user(self) -> str:
        return (self.settings_service.get_setting(SETTING_KEY_USER, "") or "").strip()

    def issue_type(self) -> str:
        return (
            self.settings_service.get_setting(SETTING_KEY_ISSUE_TYPE, DEFAULT_JIRA_ISSUE_TYPE)
            or DEFAULT_JIRA_ISSUE_TYPE
        ).strip()

    def labels(self) -> List[str]:
        raw = self.settings_service.get_setting(SETTING_KEY_LABELS, DEFAULT_LABELS) or ""
        return [label.strip() for label in raw.split(",") if label.strip()]

    def allow_private_url(self) -> bool:
        return self.settings_service.get_setting(SETTING_KEY_ALLOW_PRIVATE, "true") == "true"

    def token(self) -> str:
        """The decrypted token, or an empty string.

        Never raises: a token that cannot be decrypted — a rotated `ENCRYPTION_KEY`,
        say — must disable ticket creation, not break the scheduler tick that calls
        this.
        """
        stored = (self.settings_service.get_setting(SETTING_KEY_TOKEN, "") or "").strip()
        if not stored or not self.encryption_service:
            return stored
        try:
            return self.encryption_service.decrypt(stored, context=TOKEN_CONTEXT)
        except Exception:
            logger.error("Ticket token could not be decrypted — ticket creation disabled")
            return ""

    def set_token(self, raw_token: str) -> None:
        """Store the token encrypted, bound to its own setting key."""
        raw_token = (raw_token or "").strip()
        if not raw_token:
            self.settings_service.update_setting(SETTING_KEY_TOKEN, "")
            return
        if not self.encryption_service:
            raise RuntimeError("Le service de chiffrement est requis pour enregistrer un jeton.")
        encrypted = self.encryption_service.encrypt(raw_token, context=TOKEN_CONTEXT)
        self.settings_service.update_setting(SETTING_KEY_TOKEN, encrypted)

    def set_base_url(self, url: str) -> None:
        """Validated at save time, and again before every request."""
        self.settings_service.update_setting(
            SETTING_KEY_BASE_URL, self._validated_base_url(url)
        )

    def _validated_base_url(self, url: str) -> str:
        return validate_outbound_url(
            url, allow_private=self.allow_private_url(), label="URL du gestionnaire de tickets"
        ).rstrip("/")

    # --- Creating ---

    def create_for_issue(self, issue: Issue, target_name: str) -> Optional[Ticket]:
        """Open one ticket. Returns `None` on any failure, having logged it.

        Never raises: this runs from the scheduler tick, and an unreachable tracker
        must not stop the retention pass or the triage expiry that share it.
        """
        if not self.is_enabled():
            return None
        try:
            base_url = self._validated_base_url(self.base_url())
        except UnsafeUrlError as e:
            logger.error("Ticket not created: %s", e)
            return None

        provider = self.provider()
        title = build_title(issue, target_name)
        body = build_body(issue, target_name)
        try:
            if provider == PROVIDER_GITLAB:
                return self._create_gitlab(base_url, title, body)
            return self._create_jira(base_url, title, body)
        except Exception:
            logger.exception(
                "Ticket creation failed for issue %s (non-fatal, will be retried)", issue.id
            )
            return None

    def _create_gitlab(self, base_url: str, title: str, body: str) -> Optional[Ticket]:
        # The project id must be URL-encoded when it is a path ("group/project"),
        # which is how most people have it to hand.
        from urllib.parse import quote

        url = f"{base_url}/api/v4/projects/{quote(self.project(), safe='')}/issues"
        response = self._http_post(
            url,
            headers={"PRIVATE-TOKEN": self.token()},
            json={"title": title, "description": body, "labels": ",".join(self.labels())},
            timeout=HTTP_TIMEOUT_SECONDS,
        )
        response.raise_for_status()
        payload = response.json()
        reference = f"#{payload.get('iid')}"
        return Ticket(reference=reference, url=payload.get("web_url") or "")

    def _create_jira(self, base_url: str, title: str, body: str) -> Optional[Ticket]:
        url = f"{base_url}/rest/api/3/issue"
        fields: Dict[str, Any] = {
            "project": {"key": self.project()},
            "summary": title,
            "issuetype": {"name": self.issue_type()},
            # Atlassian Document Format: Jira Cloud's v3 API refuses plain strings
            # for `description`.
            "description": {
                "type": "doc",
                "version": 1,
                "content": [
                    {"type": "paragraph", "content": [{"type": "text", "text": line}]}
                    for line in body.split("\n")
                    if line.strip()
                ],
            },
        }
        if self.labels():
            fields["labels"] = self.labels()

        response = self._http_post(
            url,
            auth=(self.user(), self.token()) if self.user() else None,
            headers={"Accept": "application/json"},
            json={"fields": fields},
            timeout=HTTP_TIMEOUT_SECONDS,
        )
        response.raise_for_status()
        payload = response.json()
        key = payload.get("key") or ""
        return Ticket(reference=key, url=f"{base_url}/browse/{key}" if key else "")


def build_title(issue: Issue, target_name: str) -> str:
    """Short enough for a tracker list, specific enough to be searchable."""
    subject = issue.identifier or issue.type
    package = f" — {issue.package_name}" if issue.package_name else ""
    severity = (issue.severity or "unknown").upper()
    return f"[Zanshin][{severity}] {subject}{package} ({target_name})"


def build_body(issue: Issue, target_name: str) -> str:
    """What the ticket says, written for whoever picks it up cold.

    The fixed version comes first among the details, because it is the difference
    between a ticket somebody can close today and a ticket that gets carried between
    three sprints.
    """
    lines = [
        f"Détecté par Zanshin sur **{target_name}**.",
        "",
        f"- Type : {issue.type}",
        f"- Identifiant : {issue.identifier or '—'}",
        f"- Sévérité : {(issue.severity or 'unknown')}",
    ]
    if issue.fix_versions:
        lines.append(f"- **Corrigé dans : {issue.fix_versions}**")
    elif issue.fix_state in ("not-fixed", "wont-fix"):
        lines.append("- Aucun correctif publié à ce jour")
    if issue.package_name:
        version = f" {issue.package_version}" if issue.package_version else ""
        lines.append(f"- Composant : {issue.package_name}{version}")
    if issue.is_direct_dependency is not None:
        lines.append(
            "- Dépendance : "
            + ("directe (déclarée par le projet)" if issue.is_direct_dependency else "transitive")
        )
    if issue.file_path:
        location = issue.file_path + (f":{issue.line}" if issue.line else "")
        lines.append(f"- Emplacement : {location}")
    if issue.is_kev:
        lines.append("- ⚠️ Exploitation active connue (catalogue CISA KEV)")
    if issue.epss_score is not None:
        lines.append(f"- Probabilité d'exploitation (EPSS) : {issue.epss_score:.1%}")
    if issue.link:
        lines.append(f"- Référence : {issue.link}")
    if issue.description:
        lines += ["", issue.description[:1000]]
    lines += [
        "",
        f"Problème Zanshin #{issue.id} — empreinte `{issue.fingerprint}`.",
        "Ce ticket a été ouvert parce que ce problème ferait échouer un build selon la "
        "politique de gate en vigueur pour cette cible.",
    ]
    return "\n".join(lines)


def sweep(
    db,
    *,
    issue_repository,
    gate_policy_service,
    ticket_service: "TicketService",
    audit_log_service=None,
    limit: int = MAX_TICKETS_PER_SWEEP,
) -> List[Issue]:
    """Open tickets for issues that would fail their target's gate. Returns those done.

    Takes its collaborators as arguments rather than holding them, for the same reason
    `IssueService` does: this runs on the scheduler's session, and a service that
    captured a session would be wrong on the second tick.

    The gate is evaluated **one issue at a time**, deliberately. Evaluating a target's
    whole backlog and ticketing every violation would be the same query, but it would
    also mean an issue's ticket depends on which other issues happen to be open — and
    "why did this one get a ticket and that one not" must have an answer that is about
    the issue itself.
    """
    if not ticket_service.is_enabled():
        return []

    from zanshin.services.policy_gate import evaluate

    created: List[Issue] = []
    for issue in issue_repository.find_actionable_without_ticket(limit=limit):
        # An issue always belongs to a target in practice, but a row without one must
        # fall back to the global policy rather than resolve a scope of
        # `("container", None)` — which would raise and take the whole scheduler tick
        # with it, retention and triage expiry included.
        if issue.repo_id:
            kind, target_id = "repository", issue.repo_id
        elif issue.container_id:
            kind, target_id = "container", issue.container_id
        else:
            kind, target_id = None, None
        resolved = gate_policy_service.resolve(kind, target_id)

        if evaluate([issue], resolved.policy).passed:
            # Below the bar for this target: no ticket, and no marker either — the
            # policy may be tightened tomorrow, and the issue should become a
            # candidate again when it is.
            continue

        ticket = ticket_service.create_for_issue(issue, _target_name(issue))
        if not ticket:
            # Left without a reference on purpose, so the next tick retries it.
            continue

        issue.ticket_ref = ticket.reference
        issue.ticket_url = ticket.url
        db.commit()
        created.append(issue)

        if audit_log_service:
            audit_log_service.record(
                "TICKET_CREATED",
                resource_id=str(issue.id),
                description=(
                    f"Ticket {ticket.reference} ouvert pour "
                    f"{issue.identifier or issue.type} ({resolved.description})"
                ),
            )

    if created:
        logger.info("Opened %d tracker ticket(s)", len(created))
    return created


def _target_name(issue: Issue) -> str:
    if issue.repository is not None:
        return issue.repository.name or issue.repository.url
    if issue.container is not None:
        return issue.container.image_string
    return "cible inconnue"

File: vectispire-java/vectispire-core/src/main/java/com/asmolabs/vectispire/core/services/TicketService.java
package com.asmolabs.vectispire.core.services;

import com.asmolabs.vectispire.common.domain.crypto.SecretCipher;
import com.asmolabs.vectispire.common.domain.net.OutboundPolicy;
import com.asmolabs.vectispire.common.domain.settings.Setting;
import com.asmolabs.vectispire.common.domain.tickets.TicketProvider;
import com.asmolabs.vectispire.common.domain.tickets.Tickets;
import com.asmolabs.vectispire.common.domain.tickets.Tickets.TicketableIssue;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Base64;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

/**
 * Opening a ticket with GitLab or Jira.
 *
 * <p><b>The token is encrypted at rest.</b> It grants write access to the tracker, which is a
 * different class of secret from a webhook URL: it therefore goes through the encryption
 * service like an SSH key, instead of sitting in the clear in the settings table.
 *
 * <p><b>One ticket per issue, for its whole life.</b> Not per scan, and never reopened when the
 * issue comes back: a ticket that rises from the dead on every rescan is how people learn to
 * mute a project's notifications. The stored reference is set once and never cleared — it is
 * also the sweep's deduplication key.
 */
@Service
public class TicketService {

    private static final Logger log = LoggerFactory.getLogger(TicketService.class);

    /** The token's encryption context, bound to its setting key. */
    public static final String TOKEN_CONTEXT = "setting:ticket_token";

    private final SettingsService settings;
    private final EncryptionService encryption;
    private final OutboundPost post;
    private final ObjectMapper json;

    public TicketService(
            SettingsService settings, EncryptionService encryption, OutboundPost post, ObjectMapper json) {
        this.settings = settings;
        this.encryption = encryption;
        this.post = post;
        this.json = json;
    }

    /** @param reference what the tracker calls it; @param url where a human goes to read it */
    public record Ticket(String reference, String url) {}

    public TicketProvider provider() {
        return TicketProvider.fromWireName(settings.get(Setting.TICKET_PROVIDER)).orElse(TicketProvider.NONE);
    }

    public String baseUrl() {
        return trimTrailingSlashes(settings.get(Setting.TICKET_BASE_URL).trim());
    }

    public String project() {
        return settings.get(Setting.TICKET_PROJECT).trim();
    }

    public boolean isEnabled() {
        return provider().isEnabled() && !baseUrl().isEmpty() && !project().isEmpty() && !token().isEmpty();
    }

    /**
     * The decrypted token, or an empty string.
     *
     * <p><b>Never throws</b>: an undecryptable token — a rotated encryption key, say — must
     * disable ticket creation, not break the maintenance tick that calls this and that also
     * carries the purge and triage expiry.
     */
    public String token() {
        String stored = settings.get(Setting.TICKET_TOKEN).trim();
        if (stored.isEmpty()) {
            return stored;
        }

        // Tolerates a token written before this setting was encrypted — reported, never silent.
        // See EncryptionService.readSecret for why refusing it outright would disable the
        // integration rather than protect the row.
        return encryption.readSecret(stored, TOKEN_CONTEXT, "The tracker access token");
    }

    /** Binds the inbound webhook secret to its own row, like the token above. */
    public static final String WEBHOOK_SECRET_CONTEXT = "setting:ticket_webhook_secret";

    /**
     * The secret the tracker presents when it calls us, decrypted.
     *
     * <p>Read through the same tolerant path as the token: this one authenticates <b>the only
     * anonymous mutating route in the system</b>, so a value that stops being readable does not
     * fail closed in a way anybody notices — it makes the route refuse the real tracker, and a
     * triage decision simply stops arriving.
     */
    public String webhookSecret() {
        return encryption.readSecret(
                settings.get(Setting.TICKET_WEBHOOK_SECRET).trim(),
                WEBHOOK_SECRET_CONTEXT,
                "The inbound webhook secret");
    }

    /** Whether one is configured, for a screen that must not show it. */
    public boolean hasWebhookSecret() {
        return !settings.get(Setting.TICKET_WEBHOOK_SECRET).trim().isEmpty();
    }

    /** Stores it encrypted. Blank clears it, which leaves the webhook route anonymous. */
    public void setWebhookSecret(String rawSecret) {
        String value = rawSecret == null ? "" : rawSecret.trim();
        settings.set(
                Setting.TICKET_WEBHOOK_SECRET,
                value.isEmpty() ? "" : encryption.encrypt(value, WEBHOOK_SECRET_CONTEXT));
    }

    /** Stores the token encrypted, bound to its own setting key. */
    public void setToken(String rawToken) {
        String value = rawToken == null ? "" : rawToken.trim();
        settings.set(Setting.TICKET_TOKEN, value.isEmpty() ? "" : encryption.encrypt(value, TOKEN_CONTEXT));
    }

    /**
     * Opens a ticket. Empty on any failure, after logging it.
     *
     * <p>Never throws: this runs from the maintenance tick, and an unreachable tracker must not
     * stop the other jobs that share it.
     */
    public Optional<Ticket> createForIssue(TicketableIssue issue, String targetName) {
        if (!isEnabled()) {
            return Optional.empty();
        }

        String baseUrl;
        try {
            // **Validated here too**, and not only when saved: a setting written straight into
            // the database must not become an unchecked destination.
            baseUrl = validatedBaseUrl(baseUrl());
        } catch (RuntimeException refused) {
            log.error("Ticket not created: {}", refused.getMessage());
            return Optional.empty();
        }

        try {
            String title = Tickets.title(issue, targetName);
            String body = Tickets.body(issue, targetName);
            return switch (provider()) {
                case GITLAB -> Optional.of(createGitlab(baseUrl, title, body));
                case GITHUB -> Optional.of(createGithub(baseUrl, title, body));
                case JIRA -> Optional.of(createJira(baseUrl, title, body));
                case SERVICENOW -> Optional.of(createServiceNow(baseUrl, title, body));
                case NONE -> Optional.empty();
            };
        } catch (RuntimeException failed) {
            log.warn("Ticket creation failed for issue {} — will be retried: {}", issue.id(), failed.getMessage());
            return Optional.empty();
        }
    }

    /**
     * Closes an existing ticket when an issue is verified resolved.
     */
    public boolean closeTicket(String ticketRef, String resolutionReason) {
        if (!isEnabled() || ticketRef == null || ticketRef.isBlank()) {
            return false;
        }

        String baseUrl;
        try {
            baseUrl = validatedBaseUrl(baseUrl());
        } catch (RuntimeException refused) {
            log.error("Ticket closure failed - invalid base URL: {}", refused.getMessage());
            return false;
        }

        try {
            switch (provider()) {
                case GITLAB -> closeGitlab(baseUrl, ticketRef, resolutionReason);
                case GITHUB -> closeGithub(baseUrl, ticketRef, resolutionReason);
                case JIRA -> closeJira(baseUrl, ticketRef, resolutionReason);
                case SERVICENOW -> closeServiceNow(baseUrl, ticketRef, resolutionReason);
                case NONE -> {}
            }
            log.info("Successfully closed ticket {} on {}", ticketRef, provider());
            return true;
        } catch (RuntimeException e) {
            log.warn("Failed to close ticket {} on {}: {}", ticketRef, provider(), e.getMessage());
            return false;
        }
    }

    /**
     * The base URL, validated.
     *
     * <p><b>Private is allowed by default here</b>, unlike the notification webhook: a
     * self-hosted GitLab or Jira commonly lives on an internal network. The setting remains, so
     * a deployment that only uses a hosted tracker can forbid it.
     */
    public String validatedBaseUrl(String url) {
        return trimTrailingSlashes(post.validate(url, policy(), "tracker URL"));
    }

    /**
     * The same policy for the base URL and for the call that follows it.
     *
     * <p>Read once and passed on rather than recomputed at each site: a base URL validated
     * under one policy and posted to under another would mean the check that ran is not the
     * check that applies.
     */
    private OutboundPolicy policy() {
        return settings.isEnabled(Setting.TICKET_ALLOW_PRIVATE_URL)
                ? OutboundPolicy.INTERNAL_ALLOWED
                : OutboundPolicy.PUBLIC_ONLY;
    }

    private List<String> labels() {
        return Tickets.parseLabels(settings.get(Setting.TICKET_LABELS));
    }

    private Ticket createGitlab(String baseUrl, String title, String body) {
        // The project identifier has to be encoded when it is a path ("group/project"), which is
        // the form most people have it in.
        String url = baseUrl + "/api/v4/projects/" + URLEncoder.encode(project(), StandardCharsets.UTF_8) + "/issues";
        JsonNode payload = read(post.postForResponse(
                url,
                Map.of("title", title, "description", body, "labels", String.join(",", labels())),
                policy(),
                "GitLab",
                Map.of("PRIVATE-TOKEN", token())));

        return new Ticket("#" + payload.path("iid").asText(""), payload.path("web_url").asText(""));
    }

    private void closeGitlab(String baseUrl, String ticketRef, String resolutionReason) {
        String iid = ticketRef.startsWith("#") ? ticketRef.substring(1) : ticketRef;
        String url = baseUrl + "/api/v4/projects/" + URLEncoder.encode(project(), StandardCharsets.UTF_8) + "/issues/" + iid;
        post.postForResponse(url, Map.of("state_event", "close"), policy(), "GitLab", Map.of("PRIVATE-TOKEN", token()));
    }

    private Ticket createGithub(String baseUrl, String title, String body) {
        String url = baseUrl + "/repos/" + project() + "/issues";
        Map<String, Object> payload = new java.util.LinkedHashMap<>();
        payload.put("title", title);
        payload.put("body", body);
        if (!labels().isEmpty()) {
            payload.put("labels", labels());
        }

        JsonNode response = read(post.postForResponse(
                url,
                payload,
                policy(),
                "GitHub",
                Map.of("Authorization", "Bearer " + token(), "Accept", "application/vnd.github+json")));

        return new Ticket("#" + response.path("number").asText(""), response.path("html_url").asText(""));
    }

    private void closeGithub(String baseUrl, String ticketRef, String resolutionReason) {
        String number = ticketRef.startsWith("#") ? ticketRef.substring(1) : ticketRef;
        String url = baseUrl + "/repos/" + project() + "/issues/" + number;
        post.postForResponse(
                url,
                Map.of("state", "closed", "state_reason", "completed"),
                policy(),
                "GitHub",
                Map.of("Authorization", "Bearer " + token(), "Accept", "application/vnd.github+json"));
    }

    private Ticket createJira(String baseUrl, String title, String body) {
        Map<String, Object> fields = new java.util.LinkedHashMap<>();
        fields.put("project", Map.of("key", project()));
        fields.put("summary", title);
        fields.put("issuetype", Map.of("name", settings.get(Setting.TICKET_ISSUE_TYPE)));
        // Atlassian Document Format: Jira Cloud's v3 API refuses a plain string for `description`.
        fields.put("description", atlassianDocument(body));
        if (!labels().isEmpty()) {
            fields.put("labels", labels());
        }

        JsonNode payload = read(post.postForResponse(
                baseUrl + "/rest/api/3/issue",
                Map.of("fields", fields),
                policy(),
                "Jira",
                jiraHeaders()));

        String key = payload.path("key").asText("");
        return new Ticket(key, key.isEmpty() ? "" : baseUrl + "/browse/" + key);
    }

    private void closeJira(String baseUrl, String ticketRef, String resolutionReason) {
        String url = baseUrl + "/rest/api/3/issue/" + ticketRef + "/transitions";
        // Default Jira Cloud transition or comment
        post.postForResponse(
                url,
                Map.of("transition", Map.of("id", "31")),
                policy(),
                "Jira",
                jiraHeaders());
    }

    private Ticket createServiceNow(String baseUrl, String title, String body) {
        String url = baseUrl + "/api/now/table/incident";
        Map<String, Object> payload = new java.util.LinkedHashMap<>();
        payload.put("short_description", title);
        payload.put("description", body);
        payload.put("category", "Security");
        payload.put("urgency", "1");
        payload.put("impact", "1");

        JsonNode response = read(post.postForResponse(url, payload, policy(), "ServiceNow", serviceNowHeaders()));
        JsonNode result = response.path("result");
        String number = result.path("number").asText("");
        String sysId = result.path("sys_id").asText("");
        String webUrl = baseUrl + "/nav_to.do?uri=incident.do?sys_id=" + (sysId.isEmpty() ? number : sysId);
        return new Ticket(number.isEmpty() ? sysId : number, webUrl);
    }

    private void closeServiceNow(String baseUrl, String ticketRef, String resolutionReason) {
        String url = baseUrl + "/api/now/table/incident/" + ticketRef;
        post.postForResponse(
                url,
                Map.of("state", "6", "close_code", "Solved (Permanently)", "close_notes", "Resolved by Vectispire: " + resolutionReason),
                policy(),
                "ServiceNow",
                serviceNowHeaders());
    }

    private Map<String, String> jiraHeaders() {
        Map<String, String> headers = new java.util.LinkedHashMap<>();
        headers.put("Accept", "application/json");
        String user = settings.get(Setting.TICKET_USER).trim();
        if (!user.isEmpty()) {
            String credentials = Base64.getEncoder()
                    .encodeToString((user + ":" + token()).getBytes(StandardCharsets.UTF_8));
            headers.put("Authorization", "Basic " + credentials);
        }
        return headers;
    }

    private Map<String, String> serviceNowHeaders() {
        Map<String, String> headers = new java.util.LinkedHashMap<>();
        headers.put("Accept", "application/json");
        String user = settings.get(Setting.TICKET_USER).trim();
        if (!user.isEmpty()) {
            String credentials = Base64.getEncoder()
                    .encodeToString((user + ":" + token()).getBytes(StandardCharsets.UTF_8));
            headers.put("Authorization", "Basic " + credentials);
        } else {
            headers.put("Authorization", "Bearer " + token());
        }
        return headers;
    }

    private static Map<String, Object> atlassianDocument(String body) {
        List<Object> paragraphs = new ArrayList<>();
        for (String line : body.split("\n")) {
            if (!line.isBlank()) {
                paragraphs.add(Map.of("type", "paragraph", "content", List.of(Map.of("type", "text", "text", line))));
            }
        }
        return Map.of("type", "doc", "version", 1, "content", paragraphs);
    }

    private JsonNode read(String response) {
        try {
            return json.readTree(response);
        } catch (JsonProcessingException notJson) {
            throw new IllegalStateException("The tracker answered with something that is not JSON", notJson);
        }
    }

    private static String trimTrailingSlashes(String url) {
        return url.replaceAll("/+$", "");
    }
}
